src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php

```diff
@@ -0,0 +1,281 @@
+                            $cartRules[$orderInvoice->id]['value_tax_excl']
+                        );
+                    }
+                } else {
+                    if ($discountValue > $order->total_paid_tax_incl) {
+                        throw new OrderException('The discount value is greater than the order total.');
+                    }
+
+                    $cartRules[0]['value_tax_incl'] = Tools::ps_round($discountValue, 2);
+                    $cartRules[0]['value_tax_excl'] = Tools::ps_round(
+                        $discountValue / (1 + ($order->getTaxesAverageUsed() / 100)),
+                        2
+                    );
+                }
+
+                break;
+            // Free shipping type
+            case OrderDiscountType::FREE_SHIPPING:
+                if (isset($orderInvoice)) {
+                    if ($orderInvoice->total_shipping_tax_incl > 0) {
+                        $cartRules[$orderInvoice->id]['value_tax_incl'] = $orderInvoice->total_shipping_tax_incl;
+                        $cartRules[$orderInvoice->id]['value_tax_excl'] = $orderInvoice->total_shipping_tax_excl;
+
+                        // Update OrderInvoice
+                        $this->applyDiscountOnInvoice(
+                            $orderInvoice,
+                            $cartRules[$orderInvoice->id]['value_tax_incl'],
+                            $cartRules[$orderInvoice->id]['value_tax_excl']
+                        );
+                    }
+                } elseif ($order->hasInvoice()) {
+                    $orderInvoices = $order->getInvoicesCollection();
+                    foreach ($orderInvoices as $orderInvoice) {
+                        /** @var OrderInvoice $orderInvoice */
+                        if ($orderInvoice->total_shipping_tax_incl <= 0) {
+                            continue;
+                        }
+                        $cartRules[$orderInvoice->id]['value_tax_incl'] = $orderInvoice->total_shipping_tax_incl;
+                        $cartRules[$orderInvoice->id]['value_tax_excl'] = $orderInvoice->total_shipping_tax_excl;
+
+                        // Update OrderInvoice
+                        $this->applyDiscountOnInvoice(
+                            $orderInvoice,
+                            $cartRules[$orderInvoice->id]['value_tax_incl'],
+                            $cartRules[$orderInvoice->id]['value_tax_excl']
+                        );
+                    }
+                } else {
+                    $cartRules[0]['value_tax_incl'] = $order->total_shipping_tax_incl;
+                    $cartRules[0]['value_tax_excl'] = $order->total_shipping_tax_excl;
+                }
+
+                break;
+            default:
+               throw new OrderException('The discount type is invalid.');
+        }
+
+        $res = true;
+        foreach ($cartRules as &$cartRule) {
+            $cartRuleObj = new CartRule();
```

Comment:
I would like to improve this although this is not the scope of this PR.
Do you think we can at least add `@todo` to indicate areas of improvements ?

Here we could create a `private` function to create the proper CartRuleObject.